src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php

```diff
@@ -161,46 +148,76 @@ protected function doDelete(array $ids, array $tagData = []): bool
      */
     protected function doInvalidate(array $tagIds): bool
     {
-        if (!$this->redisServerSupportSPOP()) {
+        if (!$this->redis instanceof \Predis\ClientInterface || !$this->redis->getConnection() instanceof PredisCluster) {
+            $movedTagSetIds = $this->renameKeys($this->redis, $tagIds);
+        } else {
+            $clusterConnection = $this->redis->getConnection();
+            $tagIdsByConnection = new \SplObjectStorage();
+            $movedTagSetIds = [];
+
+            foreach ($tagIds as $id) {
+                $connection = $clusterConnection->getConnectionByKey($id);
+                $slot = $tagIdsByConnection[$connection] ?? $tagIdsByConnection[$connection] = new \ArrayObject();
+                $slot[] = $id;
+            }
+
+            foreach ($tagIdsByConnection as $connection) {
+                $slot = $tagIdsByConnection[$connection];
+                $movedTagSetIds = array_merge($movedTagSetIds, $this->renameKeys(new $this->redis($connection, $this->redis->getOptions()), $slot->getArrayCopy()));
```

Comment:
should we actually use a dynamic class name coming from the instance here, or do `new Predis\Client()` instead (as we rely on the constructor signature of the class anyway)

Reply:
Internally, Predis does this also, so we're consistent here.